File: src/slate/lexer.zig
//:___________________________________________________________________
//  *Slate  |  Copyright (C) Ivan Mar (sOkam!)  |  LGPLv3 or higher  :
//:___________________________________________________________________
//! @fileoverview
//!  Contains a Generic Lexer.
//!  Its process will create a list of lexemes,
//!  that will be the input for the Tokenizer process of the target language.
//! @note
//!  A lexeme is a sequence of characters in the source program
//!  that matches the pattern for a token
//!  and is identified by the lexical analyzer as an instance of that token.
//___________________________________________________________________________|
pub const Lex = @This();
// @deps std
const std = @import("std");
// @deps zstd
const zstd = @import("../zstd.zig");
// @deps *Slate
const source  = @import("./source.zig").source;
const Ch      = @import("./char.zig");
const Lx      = @import("./lex/lexeme.zig");
pub const Loc = source.Loc;
pub const Pos = source.Pos;


//______________________________________
/// @descr
/// Describes a Lexer process and its data.
/// @in A sequence of ascii characters  (cstr)
/// @out The list of lexemes that those characters represent.
//____________________________
/// @field {@link Lex.A} The Allocator used by the Lexer
A    :std.mem.Allocator,
/// @field {@link Lex.pos} The current {@link Lex.src} position read by the Lexer.
pos  :source.Pos,
/// @field {@link Lex.src} The sequence of ascii characters that are being lexed.
src  :source.Code,
/// @field {@link Lex.res} The list of lexemes resulting from the Lexer process.
res  :Lx.List,


//__________________________
/// @descr Creates a new Lexer object from the given {@arg src} code data.
pub fn create (A :std.mem.Allocator, src :source.Code) !Lex {
  return Lex{
    .A   = A,
    .pos = 0,
    .src = src,
    .res = Lx.List{},
    };
}

//__________________________
/// @descr Frees all resources owned by the Lexer object.
pub fn destroy (L:*Lex) void {
  L.res.deinit(L.A);
}

//__________________________
pub fn cloneResult (L:*Lex) !Lx.List {
  return L.res.clone(L.A);
}
//______________________________________
// @section Lexer: General Tools
//__________________________
pub const report = @import("./lex/cli.zig").report;
pub const prnt   = zstd.prnt;
pub fn fail(err :anyerror, comptime msg :zstd.cstr, args :anytype) !void {
  if (std.debug.runtime_safety) zstd.prnt("[slate.lexer.debug] Err: " ++ msg ++ "\n", args);
  return err;
}


//______________________________________
// @section Lexer: State/Data Management
pub const data       = @import("./lex/data.zig");
pub const add        = data.add.one;
pub const add_toLast = data.add.toLast;
pub const add_single = data.add.single;
pub const ch         = data.ch;


//______________________________________
// @section Lexer Process: Ident+Literals
pub const others     = @import("./lex/others.zig");
pub const ident      = others.ident;
pub const number     = others.number;
//______________________________________
// @section Lexer Process: Symbols
pub const symbols    = @import("./lex/symbols.zig");
pub const paren      = symbols.paren;
pub const brace      = symbols.brace;
pub const bracket    = symbols.bracket;
pub const eq         = symbols.eq;
pub const at         = symbols.at;
pub const star       = symbols.star;
pub const colon      = symbols.colon;
pub const semicolon  = symbols.semicolon;
pub const dot        = symbols.dot;
pub const comma      = symbols.comma;
pub const hash       = symbols.hash;
pub const quote_S    = symbols.quote_S;
pub const quote_D    = symbols.quote_D;
pub const quote_B    = symbols.quote_B;
pub const slash_F    = symbols.slash_F;
pub const slash_B    = symbols.slash_B;
//______________________________________
// @section Lexer Process: Whitespace
pub const whitespace = @import("./lex/whitespace.zig");
pub const newline    = whitespace.newline;
pub const space      = whitespace.space;
//__________________________
/// @descr Lexer Process: Entry Point
pub fn process(L:*Lex) !void {
  while (L.pos < L.src.len) : (L.pos += 1) {
    const c = L.ch();
    switch (c) {
    'a'...'z',
    'A'...'Z',
    '_',       => try L.ident(),
    '0'...'9'  => try L.number(),
    '*'        => try L.star(),
    '(', ')'   => try L.paren(),
    '{', '}'   => try L.brace(),
    '[', ']'   => try L.bracket(),
    ':'        => try L.colon(),
    ';'        => try L.semicolon(),
    '.'        => try L.dot(),
    ','        => try L.comma(),
    '='        => try L.eq(),
    '@'        => try L.at(),
    '#'        => try L.hash(),
    '\''       => try L.quote_S(),
    '\"'       => try L.quote_D(),
    '`'        => try L.quote_B(),
    '/'        => try L.slash_F(),
    '\\'       => try L.slash_B(),
    ' '        => try L.space(),
    '\n'       => try L.newline(),
    else => |char| try Lex.fail(error.slate_lexer_UnknownFirstCharacter, "Unknown first character '{c}' (0x{X})", .{char, char})
    }
  }
}

File: src/nim/slate/lexer.nim
#:___________________________________________________________________
#  *Slate  |  Copyright (C) Ivan Mar (sOkam!)  |  LGPLv3 or higher  :
#:___________________________________________________________________
# @deps std
from std/strformat import `&`
from std/strutils  import `%`, join
# @deps ndk
import ./nstd
# @deps Lex
import ./chars
import ./source


#_______________________________________
# @section Lexeme
#_____________________________
## @descr {@link Lx.id} Valid kinds for Lexemes
type Id *{.pure.}= enum
  ident,
  number,
  underscore,  # _
  colon,       # :
  equal,       # =
  star,        # *
  paren_L,     # (
  paren_R,     # )
  hash,        # #
  semicolon,   # ;
  quote_S,     # '  (single quote)
  quote_D,     # "  (double quote)
  quote_B,     # `  (backtick quote)
  brace_L,     # {
  brace_R,     # }
  bracket_L,   # [
  bracket_R,   # ]
  dot,         # .
  comma,       # ,
  # Operators
  plus,        # +
  minus,       # -  (also dash, min, minus)
  slash_F,     # /  (forward slash)
  less,        # <
  more,        # >
  at,          # @
  dollar,      # $
  tilde,       # ~
  ampersand,   # &
  percent,     # %
  pipe,        # |
  exclamation, # !
  question,    # ?
  caret,       # ^
  slash_B,     # \  (backward slash)
  # Whitespace
  space,       # ` `
  newline,     # \n \r
  tab,         # \t
  EOF,         # 0x0  (the null character 0 is treated as an EOF marker)
type lexeme_Id  = lexer.Id
#___________________
type Lx * = object
  id   *:lexeme_Id
  loc  *:source.Loc
type List * = seq[Lx]
type lexeme_List = lexer.List
#___________________
func From *(lx :Lx; src :source.Code) :string= lx.loc.From(src)


#_______________________________________
# @section Lexer Errors
#_____________________________
type LexerError = object of CatchableError
type UnknownFirstCharacterError = object of LexerError
type InvalidFirstCharacterError = object of LexerError
template fail *(Err :typedesc[CatchableError]; msg :varargs[string, `$`])=
  ## @descr Marks a block of code as an unrecoverable fatal error. Raises an exception when entering the block.
  const inst = instantiationInfo()
  const info = "$#($#,$#): " % [inst.fileName, $inst.line, $inst.column]
  {.cast(noSideEffect).}: raise newException(Err, info & "\n " & msg.join(" "))


#_______________________________________
# @section Lexer Management
#_____________________________
type Lex * = object
  pos  *:source.Pos  = 0
  src  *:source.Code = ""
  res  *:lexeme_List = @[]
func create   *(_:typedesc[Lex]; src :source.Code) :Lex= Lex(src: src)
func destroy  *(L :var Lex) :void= L = Lex()
func pos_next *(L :Lex; pos :source.Pos) :source.Pos {.inline.}=
  result = L.pos+pos
  if result >= L.src.len.Sz: result = L.src.len-1
func last     *(L :Lex) :bool= L.pos == L.src.len.Sz-1
func next     *(L :Lex; pos :source.Pos) :char {.inline.}= L.src[L.pos_next(pos)]
func move     *(L :var Lex; pos :source.Pos) :void {.inline.}= L.pos = L.pos_next(pos)
func ch       *(L :Lex) :char {.inline.}= L.next(0)
func add      *(L :var Lex; id :lexeme_Id, loc :source.Loc) :void {.inline.}= L.res.add Lx(id: id, loc: loc)
func add      *(L :var Lex; id :lexeme_Id) :void {.inline.}= L.add id, source.Loc(start:L.pos, End:L.pos)


#_______________________________________
# @section Identifier Characters
#_____________________________
func alphabetic *(L :var Lex) :void=
  let start = L.pos
  var End   = L.pos
  while L.ch in Ch.alphabetic:
    End = L.pos
    L.move(1)
    if L.last: break # FIX: This shouldn't be here
  L.add lexeme_Id.ident, source.Loc(start:start, End:End)
  L.move(-1)
#___________________
func digit *(L :var Lex) :void=
  let start = L.pos
  var End   = L.pos
  while L.ch in Ch.digit:
    End = L.pos
    L.move(1)
    if L.last: break # FIX: This shouldn't be here
  L.add lexeme_Id.number, source.Loc(start:start, End:End)
  # L.move(-1) # FIX: Is this missing here? Found it while porting to the TS version


#_______________________________________
# @section Symbols: Single Character
#_____________________________
# Single Entry
func exclamation *(L :var Lex) :void=  L.add lexeme_Id.exclamation
func hash        *(L :var Lex) :void=  L.add lexeme_Id.hash
func dollar      *(L :var Lex) :void=  L.add lexeme_Id.dollar
func percent     *(L :var Lex) :void=  L.add lexeme_Id.percent
func ampersand   *(L :var Lex) :void=  L.add lexeme_Id.ampersand
func star        *(L :var Lex) :void=  L.add lexeme_Id.star
func plus        *(L :var Lex) :void=  L.add lexeme_Id.plus
func comma       *(L :var Lex) :void=  L.add lexeme_Id.comma
func minus       *(L :var Lex) :void=  L.add lexeme_Id.minus
func dot         *(L :var Lex) :void=  L.add lexeme_Id.dot
func colon       *(L :var Lex) :void=  L.add lexeme_Id.colon
func semicolon   *(L :var Lex) :void=  L.add lexeme_Id.semicolon
func equal       *(L :var Lex) :void=  L.add lexeme_Id.equal
func question    *(L :var Lex) :void=  L.add lexeme_Id.question
func at          *(L :var Lex) :void=  L.add lexeme_Id.at
func pipe        *(L :var Lex) :void=  L.add lexeme_Id.pipe
func tilde       *(L :var Lex) :void=  L.add lexeme_Id.tilde
func underscore  *(L :var Lex) :void=  L.add lexeme_Id.underscore
#___________________
# Groups
func whitespace *(L :var Lex) :void=
  case L.ch
  of Ch.space   : L.add lexeme_Id.space
  of Ch.tab     : L.add lexeme_Id.tab
  of Ch.newline : L.add lexeme_Id.newline
  else: lexer.fail LexerError, "whitespace  : `" & L.src[L.pos] & "`"
func quote *(L :var Lex) :void=
  case L.ch
  of Ch.quote_D : L.add lexeme_Id.quote_D
  of Ch.quote_S : L.add lexeme_Id.quote_S
  of Ch.quote_B : L.add lexeme_Id.quote_B
  else: lexer.fail LexerError, "quote: `" & L.src[L.pos] & "`"
func parenthesis *(L :var Lex) :void=
  case L.ch
  of Ch.paren_L : L.add lexeme_Id.paren_L
  of Ch.paren_R : L.add lexeme_Id.paren_R
  else: lexer.fail LexerError, "parenthesis: `" & L.src[L.pos] & "`"
func slash *(L :var Lex) :void=
  case L.ch
  of Ch.slash_B : L.add lexeme_Id.slash_B
  of Ch.slash_F : L.add lexeme_Id.slash_F
  else: lexer.fail LexerError, "slash: `" & L.src[L.pos] & "`"
func bracket *(L :var Lex) :void=
  case L.ch
  of Ch.bracket_L : L.add lexeme_Id.bracket_L
  of Ch.bracket_R : L.add lexeme_Id.bracket_R
  else: lexer.fail LexerError, "bracket: `" & L.src[L.pos] & "`"
func chevron *(L :var Lex) :void=
  case L.ch
  of Ch.less  : L.add lexeme_Id.less
  of Ch.more  : L.add lexeme_Id.more
  of Ch.caret : L.add lexeme_Id.caret
  else: lexer.fail LexerError, "chevron: `" & L.src[L.pos] & "`"
func brace *(L :var Lex) :void=
  case L.ch
  of Ch.brace_L : L.add lexeme_Id.brace_L
  of Ch.brace_R : L.add lexeme_Id.brace_R
  else: lexer.fail LexerError, "brace: `" & L.src[L.pos] & "`"



#_______________________________________
# @section Lexer: Entry Point
#_____________________________
func process *(L :var Lex) :void=
  while L.pos < L.src.len.Sz:
    case L.ch
    of Ch.whitespace  : L.whitespace
    of Ch.alphabetic  : L.alphabetic
    of Ch.digit       : L.digit
    of Ch.parenthesis : L.parenthesis
    of Ch.star        : L.star
    of Ch.colon       : L.colon
    of Ch.equal       : L.equal
    of Ch.exclamation : L.exclamation
    of Ch.quote       : L.quote
    of Ch.hash        : L.hash
    of Ch.dollar      : L.dollar
    of Ch.percent     : L.percent
    of Ch.ampersand   : L.ampersand
    of Ch.plus        : L.plus
    of Ch.comma       : L.comma
    of Ch.minus       : L.minus
    of Ch.dot         : L.dot
    of Ch.slash       : L.slash
    of Ch.semicolon   : L.semicolon
    of Ch.chevron     : L.chevron
    of Ch.question    : L.question
    of Ch.at          : L.at
    of Ch.bracket     : L.bracket
    of Ch.underscore  : L.underscore
    of Ch.brace       : L.brace
    of Ch.pipe        : L.pipe
    of Ch.tilde       : L.tilde
    of Ch.ascii_ext   : lexer.fail InvalidFirstCharacterError, &"TODO: Lexing character {$L.src[L.pos].ord}:`{L.src[L.pos]}` is not supported. Use 32..126 ascii instead."
    else              : lexer.fail UnknownFirstCharacterError, &"TODO: Lexing character {$L.src[L.pos].ord}:`{L.src[L.pos]}` not implemented."
    L.pos.inc
 
